https://bugs.gentoo.org/show_bug.cgi?id=446642

===================================================================
RCS file: /cvs/src/src/bfd/archive.c,v
retrieving revision 1.92
retrieving revision 1.93
diff -u -r1.92 -r1.93
--- src/bfd/archive.c	2012/08/20 14:32:30	1.92
+++ src/bfd/archive.c	2012/12/09 18:01:39	1.93
@@ -2424,9 +2424,6 @@
   unsigned int count;
   struct ar_hdr hdr;
   long uid, gid;
-  file_ptr max_first_real = 1;
-
-  max_first_real <<= 31;
 
   firstreal = mapsize + elength + sizeof (struct ar_hdr) + SARMAG;
 
@@ -2469,6 +2466,7 @@
 
   for (count = 0; count < orl_count; count++)
     {
+      unsigned int offset;
       bfd_byte buf[BSD_SYMDEF_SIZE];
 
       if (map[count].u.abfd != last_elt)
@@ -2488,7 +2486,8 @@
       /* The archive file format only has 4 bytes to store the offset
 	 of the member.  Check to make sure that firstreal has not grown
 	 too big.  */
-      if (firstreal >= max_first_real)
+      offset = (unsigned int) firstreal;
+      if (firstreal != (file_ptr) offset)
 	{
 	  bfd_set_error (bfd_error_file_truncated);
 	  return FALSE;
